usr/initiator.c

```diff
@@ -711,11 +711,7 @@ static void iscsi_login_eh(struct iscsi_conn *conn, struct queue_task *qtask,
 			    !iscsi_retry_initial_login(conn))
 				session_conn_shutdown(conn, qtask, err);
 			else {
-				session->reopen_cnt++;
-				session->t->template->ep_disconnect(conn);
-				if (iscsi_conn_connect(conn, qtask))
-					queue_delayed_reopen(qtask,
-						ISCSI_CONN_ERR_REOPEN_DELAY);
+			  session_conn_reopen(conn, qtask, STOP_CONN_TERM);
```

Comment:
Why do you use STOP_CONN_TERM here instead of STOP_CONN_RECOVER? Also, the indenting is incorrect for the line you've added. Lastly, your textual description of the problem is excellent, but your "fix" description is a little simplistic, as it says "don't decrement count ...", but in practice you call a function. You might add a line like "to do this, we just call session_conn_reopen() with STOP_CONN_WHATEVER, since this does the work for us.". But the comment request is just a suggestion.

Reply:
The reason for using  STOP_CONN_TERM here instead of STOP_CONN_RECOVER is that, in the kernel code, in iscsi_conn_stop() function we call iscsi_start_session_recovery(). When we enter this recovery function, we are blocking the session if the flag is STOP_CONN_RECOVER:
drivers/scsi/libiscsi.c
	if (flag == STOP_CONN_RECOVER) {
		conn->hdrdgst_en = 0;
		conn->datadgst_en = 0;
		if (session->state == ISCSI_STATE_IN_RECOVERY &&
		    old_stop_stage != STOP_CONN_RECOVER) {
			ISCSI_DBG_SESSION(session, "blocking session\n");
			iscsi_block_session(session->cls_session);
		}
	}
I think, this blocking of the session makes sense when we are in ISCSI_CONN_STATE_LOGGED_IN state for the connection. In our case, we haven't sent the login request yet. We were just in iscsi_host_set_params() in session_conn_poll() [open-iscsi code usr/initiator.c] after getting connected.

Regarding the Kernel changes, the changes are under testing for the upstream. I will follow up with the kernel patch